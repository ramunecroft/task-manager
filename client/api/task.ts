import {API_URL} from "@/config";
import {type Task} from "@/server/db/schema";

export const getTasks = async () => {
  const res = await fetch(`${API_URL}/api/task`, {
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (res.headers.get("content-type") !== "application/json") {
    return;
  }
  if (res.status === 200) {
    return res.json();
  }
  return res.json();
};

export const updateTaskStatus = async (task: Task) => {
  const res = await fetch(`${API_URL}/api/task`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title: task.title,
      description: task.description,
      ticketCode: task.ticketCode,
      status: task.status,
    }),
  });

  return res.json() as Promise<Task>;
};
