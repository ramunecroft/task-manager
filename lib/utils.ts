import {progressTriggeredAtom} from "@/store";
import {clsx, type ClassValue} from "clsx";
import {useSetAtom} from "jotai";
import {twMerge} from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatDate(date: Date, formatStr: string): string {
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, "0");
  const day = date.getDate().toString().padStart(2, "0");

  return formatStr
    .replace("yyyy", year.toString())
    .replace("MM", month)
    .replace("dd", day);
}

async function ApiFetch(
  url: string,
  options: RequestInit,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload?: any
) {
  return await fetch(url, {...options, body: JSON.stringify(payload)});
}

export {ApiFetch as fetch};
