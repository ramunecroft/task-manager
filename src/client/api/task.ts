import {API_URL} from "@/app/config";
import {type DragUpdateTaskInput} from "@/server/schema";

export const getTasks = async () => {
  const res = await fetch(`${API_URL}/api/task`, {
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (res.headers.get("content-type") !== "application/json") {
    return;
  }
  if (res.status === 200) {
    return res.json();
  }
  return res.json();
};

export const updateTaskStatus = async ({
  status,
  ticketCode,
  title,
  description,
}: DragUpdateTaskInput) => {
  const res = await fetch(`${API_URL}/api/task`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title,
      description,
      ticketCode,
      status,
    }),
  });

  return res.json();
};
