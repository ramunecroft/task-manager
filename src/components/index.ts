export * from "./search-button";
export * from "./document-modal";
