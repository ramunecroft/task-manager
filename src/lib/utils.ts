import {clsx, type ClassValue} from "clsx";
import {twMerge} from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatDate(date: Date, formatStr: string): string {
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, "0");
  const day = date.getDate().toString().padStart(2, "0");

  return formatStr
    .replace("yyyy", year.toString())
    .replace("MM", month)
    .replace("dd", day);
}

async function apiFetch<Res, Req>(
  url: string,
  options: RequestInit,
  payload?: Req
): Promise<Response> {
  const response = await fetch(url, {...options, body: JSON.stringify(payload)});
  return response;
}

export {apiFetch as fetch};
