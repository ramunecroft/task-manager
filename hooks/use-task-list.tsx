import {getTasks} from "@/client/api/task";
import {QUERY_KEY} from "@/constants";
import {type Task} from "@/server/db/schema";
import {useQuery} from "@tanstack/react-query";

export const useTaskList = () => {
  return useQuery<Task[]>({
    queryKey: [QUERY_KEY.tasklist],
    queryFn: () => getTasks(),
  });
};
