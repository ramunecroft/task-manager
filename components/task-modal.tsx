"use client";

import {TaskModalActionPannel} from "@/components/task-modal-action-pannel";
import {TaskModalHeader} from "@/components/task-modal-header";
import {TaskModalOverview} from "@/components/task-modal-overview";
import {Dialog, DialogContent} from "@/components/ui/dialog";
import {useUserList} from "@/hooks/use-user-list";
import {showTaskModalAtom} from "@/store/task";
import {useAtom} from "jotai";

export const TaskModal = () => {
  const [showTaskModal, setShowTaskModal] = useAtom(showTaskModalAtom);

  return (
    <Dialog open={showTaskModal} onOpenChange={setShowTaskModal}>
      <DialogContent
        role={"task-modal"}
        className="mx-auto my-8 flex h-3/4 max-w-md flex-col items-start rounded-lg bg-white p-4 shadow-lg sm:max-w-xl md:max-w-6xl">
        <div className="flex h-full w-full flex-col gap-y-2">
          <TaskModalHeader />
          <div id="content" className="grid h-full grid-flow-col grid-cols-6 gap-x-8">
            <TaskModalOverview />
            <TaskModalActionPannel />
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
};
