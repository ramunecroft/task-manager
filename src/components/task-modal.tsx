"use client";

import {updateTaskStatus} from "@/client/api/task";
import {TaskModalActionPannel} from "@/components/task-modal-action-pannel";
import {TaskModalHeader} from "@/components/task-modal-header";
import {TaskModalOverview} from "@/components/task-modal-overview";
import {Dialog, DialogContent} from "@/components/ui/dialog";
import {type Task} from "@/server/db/schema";
import {mutateAtom, showTaskModalAtom, taskListAtom, taskModalAtom} from "@/store/task";
import {useAtom, useAtomValue, useSetAtom} from "jotai";
import React from "react";

export const TaskModal = () => {
  const [showTaskModal, setShowTaskModal] = useAtom(showTaskModalAtom);
  const taskModal = useAtomValue(taskModalAtom);
  const taskList = useAtomValue(taskListAtom);
  const mutate = useSetAtom(mutateAtom);
  const [isOpen, setIsOpen] = React.useState(false);

  if (!taskModal) return;

  const onStatusChange = async (status: Task["status"]) => {
    await updateTaskStatus({
      ...taskModal,
      status,
    });

    const updated = taskList.map(task => {
      if (task.ticketCode === taskModal.ticketCode) {
        task.status = status;
      }
      return task;
    });
    mutate(updated);
  };

  return (
    <Dialog open={showTaskModal} onOpenChange={setShowTaskModal}>
      <DialogContent
        role={"task-modal"}
        className="mx-auto my-8 flex h-3/4 max-w-md flex-col items-start rounded-lg bg-white p-4 shadow-lg sm:max-w-xl md:max-w-6xl">
        <div className="flex h-full w-full flex-col gap-y-2">
          <TaskModalHeader task={taskModal} />
          <div id="content" className="grid h-full grid-flow-col grid-cols-6 gap-x-8">
            <TaskModalOverview />
            <TaskModalActionPannel
              isOpen={isOpen}
              setIsOpen={setIsOpen}
              selectedTask={taskModal}
              onStatusChange={onStatusChange}
            />
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
};
