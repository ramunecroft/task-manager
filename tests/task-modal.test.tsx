import {TaskModal} from "@/components/task-modal";
import {useTaskList} from "@/hooks/use-task-list";
import {showTaskModalAtom} from "@/store/task";
import {createWrapper} from "@/tests/utils";
import {
  fireEvent,
  getByTestId,
  render,
  renderHook,
  waitFor,
  screen,
  act,
} from "@testing-library/react";
import {useAtomValue, useSetAtom} from "jotai";

describe("Modal Test", () => {
  test("render task modal", async () => {
    const {result} = renderHook(() => useTaskList(), {wrapper: createWrapper()});
    const {container} = render(<TaskModal />, {wrapper: createWrapper()});

    await waitFor(() => {
      expect(result.current.isSuccess).toBe(true);
      expect(container).toMatchSnapshot();
    });
  });

  test("modal on click close modal", () => {
    const showTaskModal = renderHook(() => useAtomValue(showTaskModalAtom));
    const setShowTaskModal = renderHook(() => useSetAtom(showTaskModalAtom));
    act(() => {
      setShowTaskModal.result.current(true);
    });
    render(<TaskModal />, {wrapper: createWrapper()});
    const closeButton = screen.getByTestId("close-modal");
    fireEvent.click(closeButton);

    expect(showTaskModal.result.current).toBe(false);
  });
});
