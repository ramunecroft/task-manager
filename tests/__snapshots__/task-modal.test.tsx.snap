// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Test render task modal 1`] = `<div />`;

exports[`Modal Test renders correctly with given props 1`] = `<div />`;
